File: api/Services/InvoiceService.cs
using api.Data;
using api.Models;
using Microsoft.EntityFrameworkCore;

namespace InvoiceAPI.Services
{
	public class InvoiceService
	{
		private readonly AppDbContext _context;

		public InvoiceService(AppDbContext context)
		{
			_context = context;
		}

		public async Task<InvoiceHDR> SaveInvoiceAsync(InvoiceHDR invoice)
		{
			if (invoice.InvoiceType == "Cash")
				invoice.VAT = 0;

			decimal subtotal = invoice.Items.Sum(i => i.Quantity * i.UnitPrice);
			invoice.Total = subtotal - invoice.Discount + invoice.VAT;

			if (invoice.InvoiceType == "Credit" && invoice.Total > 10000)
				throw new Exception("Credit invoice cannot exceed 10,000 JD.");

			bool exists = await _context.Invoices.AnyAsync(i => i.InvoiceNo == invoice.InvoiceNo && i.Id != invoice.Id);
			if (exists)
				throw new Exception("Invoice number already exists.");

			if (invoice.Id == 0)
				_context.Invoices.Add(invoice);
			else
				_context.Invoices.Update(invoice);

			await _context.SaveChangesAsync();
			return invoice;
		}
	}
}

File: src/app/services/invoice-service.ts
// src/app/services/invoice.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface ItemsDTL {
  id?: number;
  invoiceId?: number;
  itemName: string;
  quantity: number;
  unitPrice: number;
  lineTotal?: number;
}

export interface InvoiceHDR {
  id?: number;
  invoiceNo: string;
  customerName: string;
  invoiceType: 'Cash' | 'Credit';
  discount: number;
  vat: number;
  total?: number;
  createdDate?: string;
  items: ItemsDTL[];
}

@Injectable({
  providedIn: 'root'
})
export class InvoiceService {
  private apiUrl = 'http://localhost:5094/api/Invoices';

  constructor(private http: HttpClient) {}

  getInvoices(): Observable<InvoiceHDR[]> {
    return this.http.get<InvoiceHDR[]>(this.apiUrl);
  }

  getInvoice(id: number): Observable<InvoiceHDR> {
    return this.http.get<InvoiceHDR>(`${this.apiUrl}/${id}`);
  }

  createInvoice(invoice: InvoiceHDR): Observable<InvoiceHDR> {
    return this.http.post<InvoiceHDR>(this.apiUrl, invoice);
  }

  updateInvoice(id: number, invoice: InvoiceHDR): Observable<InvoiceHDR> {
    return this.http.put<InvoiceHDR>(`${this.apiUrl}/${id}`, invoice);
  }

  deleteInvoice(id: number): Observable<any> {
    return this.http.delete(`${this.apiUrl}/${id}`);
  }
}
